### PR TITLE
feat: Add PyVista Qt support

### PR DESCRIPTION
**Pull request by AlejandroFernandezLuces**

### CODE REVIEW

File: .github/workflows/ci_cd.yml [hunk 1]
```diff
@@ -68,8 +69,10 @@ jobs:
   docs-build:
     name: Documentation Build
     runs-on: ubuntu-latest
-    needs: [docs-style]
+    # needs: [docs-style]
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:72]:
```suggestion
    needs: [docs-style]
```

File: .github/workflows/ci_cd.yml [hunk 2]
```diff
@@ -68,8 +69,10 @@ jobs:
...
     steps:
+    - name: Install system dependencies
+      run: sudo apt install -qq libglu1-mesa-dev libx11-xcb-dev '^libxcb*'
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:75]: Is this needed then?
```suggestion
```

File: .github/workflows/ci_cd.yml [hunk 3]
```diff
@@ -101,9 +104,11 @@ jobs:
 
   testing:
     name: Run Unit Tests
-    needs: [ smoke-tests ]
+    # needs: [ smoke-tests ]
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:107]:
```suggestion
    needs: [ smoke-tests ]
```

File: .github/workflows/ci_cd.yml [hunk 4]
```diff
@@ -101,9 +104,11 @@ jobs:
...
     runs-on: ubuntu-latest
     steps:
+      - name: Install system dependencies
+        run: sudo apt install libegl1 libxcb-cursor0 libsm6 libxext6 libxcb-xinerama0 -y
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:111]: Same here
```suggestion
```

File: pyproject.toml [hunk 5]
```diff
@@ -30,13 +30,27 @@ dependencies = [
...
+]
+
 doc = [
     "ansys-sphinx-theme==1.2.0",
     "jupyter_sphinx==0.5.3",
     "jupytext==1.16.4",
     "nbsphinx==0.9.5",
     "numpydoc==1.8.0",
+    "pyside6==6.7.3",
+    "pyvistaqt==0.11.1,<1",
```
Comment (RobPasMue) [inline pyproject.toml:53]: I would pin it to a specific version if the doc... if needed 😄 or just drop it in that case

File: src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py [hunk 6]
```diff
@@ -591,3 +600,7 @@ def plot(self, plottable_object: Any, name_filter: str = None, **plotting_option
         else:
             self.pv_interface.plot(plottable_object, name_filter, **plotting_options)
 
+    def close(self):
+        """Close the plotter for PyVistaQT."""
+        if self._use_qt:
+            self.pv_interface.scene.close()
```
Comment (RobPasMue) [inline src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py:606]: Add an EOF newline

File: src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py
Comment (RobPasMue) [inline src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py:1]: Why not have an independent pyvistaqt backend?

Comment (RobPasMue) [reply src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py:1]: Even if one extends/inherits the other..

File: src/ansys/tools/visualization_interface/backends/pyvista/pyvista_interface.py [hunk 7]
```diff
@@ -68,15 +74,27 @@ def __init__(
...
             if viz_interface.TESTING_MODE:
-                scene = pv.Plotter(off_screen=True, **plotter_kwargs)
+                if use_qt and _HAS_PYVISTAQT:
+                    scene = pyvistaqt.BackgroundPlotter(off_screen=True)
+                else:
+                    if use_qt and not _HAS_PYVISTAQT:
+                        message = "PyVistaQt dependency is not installed. Install it with " + \
+                                  "`pip install ansys-tools-visualization-interface[pyvistaqt]`."
+                        logger.warning(message)
+                    scene = pv.Plotter(off_screen=True, **plotter_kwargs)
```
Comment (RobPasMue) [inline src/ansys/tools/visualization_interface/backends/pyvista/pyvista_interface.py:91]: Same here... all this logic should go into the pyvistaqt backend implementation IMO

File: tests/test_generic_plotter.py [hunk 8]
```diff
@@ -43,6 +48,16 @@ def test_plotter_add_pd():
...
+@pytest.mark.skipif(IN_GITHUB_ACTIONS, reason="Qt breaks CICD.")
+def test_plotter_pyvistaqt():
+    """Adds polydata to the plotter."""
+    qt_backend = PyVistaBackend(use_qt=True)
+    pl = Plotter(backend=qt_backend)
+    sphere = pv.Sphere()
+    pl.plot(sphere)
+    # PyVista QT show() breaks PyTest, so we avoid it.
+    qt_backend.close()
+
```
Comment (RobPasMue) [inline tests/test_generic_plotter.py:60]: Ah you managed to test it. Then we might not need to remove the system deps you were installing, my bad

File: pyproject.toml (hunk 5 above)
Comment (AlejandroFernandezLuces) [reply pyproject.toml]: Not needed for now.. Removing it.

File: src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py
Comment (AlejandroFernandezLuces) [reply src/ansys/tools/visualization_interface/backends/pyvista/pyvista.py:1]: Refactored, I think this change implies some duplicated code, but probably has a better design. Let me know if the refactor is what you had in mind more or less.

**Comment by RobPasMue** [comment]
More or less... let's talk later. I was thinking of reusing the PyVista interface methods as much as possible by calling the parent/super ones... and even refactoring whatever is needed on the PyVista interface to make it more combinable.

**Comment by AlejandroFernandezLuces** [comment]
@RobPasMue Reverted the changes, feel free to re-review

File: .github/workflows/ci_cd.yml [hunk 9]
```diff
@@ -10,10 +10,11 @@ on:
       - main
 
 env:
-  MAIN_PYTHON_VERSION: '3.11'
+  MAIN_PYTHON_VERSION: '3.12'
   RESET_IMAGE_CACHE: 0
   PACKAGE_NAME: ansys-tools-visualization-interface
   DOCUMENTATION_CNAME: visualization-interface.tools.docs.pyansys.com
+  IN_GITHUB_ACTIONS: true
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:17]: Do we still need this?

File: tests/test_generic_plotter.py [hunk 10]
```diff
@@ -20,12 +20,17 @@
...
 from pathlib import Path
 
 import numpy as np
+import pytest
 import pyvista as pv
 
 from ansys.tools.visualization_interface import ClipPlane, MeshObjectPlot, Plotter
+from ansys.tools.visualization_interface.backends.pyvista import PyVistaBackend
+
+IN_GITHUB_ACTIONS = os.getenv("GITHUB_ACTIONS") == "true"
```
Comment (RobPasMue) [inline tests/test_generic_plotter.py:33]: Align env var naming... Looks like in your workflow this is called ``IN_GITHUB_ACTIONS`` whereas here you are trying to retrieve ``GITHUB_ACTIONS``. Does that mean that it IS being executed on CI/CD?

**Review by RobPasMue — APPROVED** [review]
LGTM - some few comments

File: tests/test_generic_plotter.py (hunk 10 above)
Comment (AlejandroFernandezLuces) [reply tests/test_generic_plotter.py:33]: Hmm.. let me test this

Comment (AlejandroFernandezLuces) [reply tests/test_generic_plotter.py]: Aligned, for some reason it was being skipped anyway. In my local tests the test is working and not being skipped.

**Comment by RobPasMue** [comment]
Feel free to merge :)

**Pull request merged**